go/src/lib.rs

```diff
@@ -250,25 +250,26 @@ pub unsafe extern "C" fn free_command_response(command_response_ptr: *mut Comman
 ///
 /// # Safety
 ///
-/// `free_error_message` can only be called once per `error_message`. Calling it twice is undefined behavior, since the address will be freed twice.
+/// `free_error_message` can only be called once per `error_message`. Calling it twice is undefined
+/// behavior, since the address will be freed twice.
 #[no_mangle]
 pub unsafe extern "C" fn free_error_message(error_message: *mut c_char) {
     assert!(!error_message.is_null());
     drop(unsafe { CString::from_raw(error_message as *mut c_char) });
 }
 
 /// Converts a double pointer to a vec.
-unsafe fn convert_double_pointer_to_vec(
-    data: *const *const c_char,
+unsafe fn convert_double_pointer_to_vec<'a>(
+    data: *const *const c_void,
     len: c_ulong,
     data_len: *const c_ulong,
-) -> Vec<Vec<u8>> {
+) -> Vec<&'a [u8]> {
     let string_ptrs = unsafe { from_raw_parts(data, len as usize) };
     let string_lengths = unsafe { from_raw_parts(data_len, len as usize) };
-    let mut result: Vec<Vec<u8>> = Vec::new();
+    let mut result = Vec::<&[u8]>::with_capacity(string_ptrs.len());
```

Comment:
We can directly use `len` in place of `string_ptrs.len()`.

Reply:
Not really. `len` is of type `u64`, `with_capacity` requires `usize`. You can't directly cast `u64` -> `usize` (possible truncation on 32bit systems)

You need to do it like this:

```rust
len.try_into().unwrap_or(usize::MAX)
```

which involves multiple calls, so we are not really using it directly

Reply:
In that case, we should use len as c_uint instead of c_ulong. I'll address this in the next PR.